refactor(LifecycleEvent): remove LifecycleEvent

vicb opened the pull request:
TODO:
- [x] dart metadata (transformer should extract data from interfaces)
- [x] doc Directive & Component should link to lifecycle hooks

vicb committed:
refactor(LifecycleEvent): remove LifecycleEvent

fixes #3924

BREAKING CHANGE

The `lifecycle` configuration for directive has been dropped.

Before

    // Dart
    @Component({lifecycle: const [LifecycleEvent.OnChanges], ...})
    class MyComponent implements OnChanges {
      void onChanges() {...}
    }

    // Typescript
    @Component({lifecycle: [LifecycleEvent.OnChanges], ...})
    class MyComponent implements OnChanges {
      onChanges(): void {...}
    }

    // ES5
    var MyComponent = ng.
    Component({lifecycle: [LifecycleEvent.OnChanges], ...}).
    Class({
      onChanges: function() {...}
    });

After

    // Dart
    @Component({...})
    class MyComponent implements OnChanges {
      void onChanges() {...}
    }

    // Typescript
    @Component({...})
    class MyComponent implements OnChanges {
      onChanges(): void {...}
    }

    // ES5
    var MyComponent = ng
      .Component({...})
      .Class({
        onChanges: function() {
        }
      });

mhevery commented on modules/angular2/src/core/compiler/interfaces.ts:
```diff
@@ -1,58 +1,72 @@
  * called after all of component's bound properties are updated.
  */
-export interface OnChanges { onChanges(changes: StringMap<string, any>): void; }
+export class OnChanges implements LifecycleHook {
```
Why did you have to make these into classes?

vicb replied:
So that we can write `hasLifeCycleHook(OnChanges, MyComponent);` and don't have to maintain a duplicate enum.

vicb commented:
@kegluneq I would appreciate some help to update the Dart transformers for this PR - I have done some updates but more is needed.

The PR is about removing the need for declaring `lifecycle: const[<list of hooks>]` and only rely on the interfaces implemented by the underlying class:
```dart
@Component({
   // lifecycle: const [LifecycleEvent.OnChanges], -> This config is dropped
   ...
})
class MyComp implements OnChanges { // -> The code now relies on the interfaces which already are the type info
  // ...
}
```

1. Could you please check if the updates to `transform/directive_processor/visitors.dart` & tests are ok,
2. `src/transform/common/directive_metadata_reader.dart` should be updated to extract the information from the implemented interfaces instead of from the `lifecycle` config.

For item 2 above, the best would be if you could send a PR to my repo: `vicb/angular`

kegluneq commented on modules/angular2/src/core/compiler/directive_lifecycle_reflector.dart:
```diff
@@ -1,36 +1,9 @@
-    return annotation.lifecycle.contains(e);
-  } else {
-    if (type is! Type) return false;
+bool hasLifecycleHook(/*LifecycleHook*/ interface, type) {
```
What is preventing us from providing types for these parameters?

kegluneq commented:
I don't see any issues with the changes so far, with the exception of your point (2) above, that `RenderDirectiveMetadata` currently won't be populated.

I would guess that is why the Travis test is failing.

I can make the updates to `directive_metadata_reader`, but I'm not certain I'll be able to do it today. I don't think there's anything particularly tricky about the change, so if you're in a hurry feel free to give it a shot.

The [current implementation](https://github.com/angular/angular/blob/master/modules_dart/transform/lib/src/transform/common/directive_metadata_reader.dart) expects to be processing either the `Directive` annotation on a class declaration or the (similar) `InstanceCreationExpression` for the same annotation that is generated for the `.ng_deps.dart` files. The problem is that `InstanceCreationExpression` does not have an associated `ClassDefinition` from which we can pull `interfaces`.

I think the easiest way to make this update is to have the reader accept, instead of the `InstanceCreationExpression` for the annotation class, the `InstanceCreationExpression` for the [`ReflectionInfo`](https://github.com/angular/angular/blob/master/modules/angular2/src/core/reflection/reflector.ts#L16) class, which specifies both the annotation(s) and the interface(s) we need.

vicb commented:
@kegluneq you're right, `directive_metadata_reader` is most probably the root cause.

You would be much more efficient than me to do the updates there. So please take a look whenever you have time.

Thanks !

Commit:
refactor(transformers): extract lifecycle hooks from the interfaces

kegluneq commented:
/cc @keertip this change required updates to `directive_metadata_reader`, which may cause some issues if it's still used to read `RenderDirectiveMetadata` in the plugin

Let me know if you see any issues stemming from this.

keertip commented:
@kegluneq , thanks for the info. Will take a look after this lands.

vicb committed:
refactor(Lifecycle hooks): move the hooks to their own module (lifecycle_hooks)

BREAKING CHANGE

Lifecycle hooks now live in the `angular2/lifecycle_hooks` module.
They previously lived in the `metadata` module.

vicb commented:
The last commit triggers a typings error, not sure how to solve this, see https://travis-ci.org/angular/angular/jobs/78499681#L260

vicb commented:
Thanks to @IgorMinar the typings generation issue has been solved.

@mhevery could you please review.

vicb committed:
fix(DirectiveResolver): Synced with latest changes

mary-poppins commented:
Merging PR #3928 on behalf of @vicb to branch [presubmit-vicb-pr-3928](https://github.com/angular/angular/tree/presubmit-vicb-pr-3928).

mary-poppins commented:
Merging PR #3928 on behalf of @vicb to branch [presubmit-vicb-pr-3928](https://github.com/angular/angular/tree/presubmit-vicb-pr-3928).

yjbanov commented:
/cc @jeffbcross

jeffbcross commented:
Thanks @yjbanov I've already (painfully) rebased against these changes :)

vicb commented:
@jeffbcross sorry for that, I know how it feels ! - I had to rebase this one >5 times to get it in :(

jeffbcross commented:
Haha @vicb no worries
